### PR TITLE
Delete backplane serviceaccounts after expiry time

### PR DESCRIPTION
**Pull request by sedroche**
Jira: https://issues.redhat.com/browse/OSD-5914

Backplane leaves serviceaccounts indefinitely atm but they should expire. This cronjob removes them after an expiry time.

### CODE REVIEW

**Comment by sedroche** [comment]
@wanghaoran1988 @karthikperu7

**Comment by wanghaoran1988** [comment]
This looks good to me, @jewzaam Could you take another look?

File: deploy/osd-delete-backplane-serviceaccounts/00-delete-backplane-serviceaccounts.rbac.yaml
Comment (jewzaam) [inline deploy/osd-delete-backplane-serviceaccounts/00-delete-backplane-serviceaccounts.rbac.yaml:32]: Should limit to backplane namespaces only.  Can use SubjectPremissions to manage this so it is based on regex.

File: deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.CronJob.yaml
Comment (jewzaam) [inline deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.CronJob.yaml:25]: When RBAC is limited to backplane namespaces I expect this needs to use the same regex to look for SA only in those same namespaces.

File: deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.CronJob.yaml
Comment (jewzaam) [inline deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.CronJob.yaml:33]: What if I'm using that SA still?  2 hours isn't that long if there's an incident.  I expect I have to log in again.  And if backplane cli were to automate that any active port forwards like Prometheus need restarted.  Backplane is proxying all requests.  Could backplane manage deletion based on the timestamp for last time the SA was used?

**Review by jewzaam — CHANGES_REQUESTED** [review]
/hold

File: deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.CronJob.yaml
Comment (wanghaoran1988) [reply deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.CronJob.yaml:33]: backplane is not caching the cluster metadata until it's being used, instead of 2 hours, maybe we can set to 12h, or 24h.

File: deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.CronJob.yaml
Comment (sedroche) [reply deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.CronJob.yaml]: SubjectPermission and filtering to openshift-backplane ns added

File: deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.CronJob.yaml
Comment (jewzaam) [reply deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.CronJob.yaml]: I like that, 12h may be a good number.  That way it should expire while I'm offline assuming a reasonable work day, even if I created the SA at the beginning or end of the work day.

Comment (jewzaam) [reply deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.CronJob.yaml]: Maybe a INFO output on ocm backplane login about when the SA expires?  it could be simple math based on the creation timestamp but requires both the cli and the job here to be on the same schedule.  small risk of drive, account for in good comments.

**Comment by sedroche** [comment]
Have updated this to 12 hours. Missed the perms for listing namespaces after the change to subjectpermissions so I have added that as well

File: deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.CronJob.yaml
Comment (sedroche) [reply deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.CronJob.yaml]: I'll create a card for this request

File: deploy/osd-delete-backplane-serviceaccounts/00-delete-backplane-serviceaccounts.rbac.yaml
Comment (jewzaam) [inline deploy/osd-delete-backplane-serviceaccounts/00-delete-backplane-serviceaccounts.rbac.yaml:58]: Not going to hold up the PR for it, but FYI you can manage global perms with SubjectPermission as well, see `.spec.clusterPermissions`.  But it does what this is doing so :+1: to what you have.

**Review by jewzaam — APPROVED** [review]

**Comment by sedroche** [comment]
@jewzaam I updated to use subjectpermssions for the clusterrolebinding. Needs one more `lgtm` to get this in

File: deploy/osd-delete-backplane-serviceaccounts/00-delete-backplane-serviceaccounts.rbac.yaml
Comment (wanghaoran1988) [inline deploy/osd-delete-backplane-serviceaccounts/00-delete-backplane-serviceaccounts.rbac.yaml:5]: This namespaces does not exists on all OSD clusters, and we should create one first, per the name, we should have a "openshift-" prefix.

File: deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.rbac.yaml
Comment (jewzaam) [inline deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.rbac.yaml:10]:
```suggestion
  name: osd-delete-backplane-serviceaccounts
```

File: deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.rbac.yaml
Comment (jewzaam) [inline deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.rbac.yaml:44]:
```suggestion
      clusterRoleName: osd-delete-backplane-serviceaccounts
```

File: deploy/osd-delete-backplane-serviceaccounts/20-delete-backplane-serviceaccounts.CronJob.yaml
Comment (jewzaam) [inline deploy/osd-delete-backplane-serviceaccounts/20-delete-backplane-serviceaccounts.CronJob.yaml:29]: Nice touch, gives us flexibility to manage in one place :+1:

File: deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.rbac.yaml
Comment (jewzaam) [inline deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.rbac.yaml:46]:
```suggestion
  subjectKind: ServiceAccount
```

**Review by jewzaam — CHANGES_REQUESTED** [review]

File: deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.rbac.yaml
Comment (sedroche) [reply deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.rbac.yaml:46]: rbac operator doesn't currently handle type of `ServiceAccount`
https://github.com/openshift/rbac-permissions-operator/blob/f0cfa43269218d756876792834ecd6c1209d8b01/pkg/controller/utils/controllerutil.go#L105-L109

The rolebinding definition for a sa is:
```yaml
subjects:
- kind: ServiceAccount
  name: osd-delete-backplane-serviceaccounts
  namespace: openshift-backplane
```

File: deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.rbac.yaml
Comment (sedroche) [reply deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.rbac.yaml]: Have update name here and in subjectpermisison

File: deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.rbac.yaml
Comment (sedroche) [reply deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.rbac.yaml:46]: @jewzaam This is the last issue on this PR. Seem ok?

Comment (jewzaam) [reply deploy/osd-delete-backplane-serviceaccounts/10-delete-backplane-serviceaccounts.rbac.yaml:46]: Yea, if User works then OK.  If not we'll know in stage.

**Comment by sedroche** [comment]
@christoph Can I get a `lgtm` for this. All feedback has been applied now

**Comment by dofinn** [comment]
Lets keep an on eye on `User` in subjectPermission CRs

**Pull request merged**
